File: internal/plugin/interfaces/network.go
package interfaces

import "context"

// NetworkProvider defines the interface for network management
type NetworkProvider interface {
	// GetName returns the name of the network provider
	GetName() string

	// GetVersion returns the version of the network provider
	GetVersion() string

	// Initialize initializes the network provider with configuration
	Initialize(config map[string]interface{}) error

	// VPC/Network Management
	CreateVPC(ctx context.Context, req CreateVPCRequest) (*VPC, error)
	GetVPC(ctx context.Context, vpcID string) (*VPC, error)
	ListVPCs(ctx context.Context) ([]VPC, error)
	UpdateVPC(ctx context.Context, vpcID string, req UpdateVPCRequest) (*VPC, error)
	DeleteVPC(ctx context.Context, vpcID string) error

	// Subnet Management
	CreateSubnet(ctx context.Context, req CreateSubnetRequest) (*Subnet, error)
	GetSubnet(ctx context.Context, subnetID string) (*Subnet, error)
	ListSubnets(ctx context.Context, vpcID string) ([]Subnet, error)
	UpdateSubnet(ctx context.Context, subnetID string, req UpdateSubnetRequest) (*Subnet, error)
	DeleteSubnet(ctx context.Context, subnetID string) error

	// Security Group Management
	CreateSecurityGroup(ctx context.Context, req CreateSecurityGroupRequest) (*SecurityGroup, error)
	GetSecurityGroup(ctx context.Context, sgID string) (*SecurityGroup, error)
	ListSecurityGroups(ctx context.Context, vpcID string) ([]SecurityGroup, error)
	UpdateSecurityGroup(ctx context.Context, sgID string, req UpdateSecurityGroupRequest) (*SecurityGroup, error)
	DeleteSecurityGroup(ctx context.Context, sgID string) error

	// Security Group Rules
	CreateSecurityGroupRule(ctx context.Context, req CreateSecurityGroupRuleRequest) (*SecurityGroupRule, error)
	GetSecurityGroupRule(ctx context.Context, ruleID string) (*SecurityGroupRule, error)
	ListSecurityGroupRules(ctx context.Context, sgID string) ([]SecurityGroupRule, error)
	DeleteSecurityGroupRule(ctx context.Context, ruleID string) error

	// Key Pair Management
	CreateKeyPair(ctx context.Context, req CreateKeyPairRequest) (*KeyPair, error)
	GetKeyPair(ctx context.Context, keyPairID string) (*KeyPair, error)
	ListKeyPairs(ctx context.Context) ([]KeyPair, error)
	DeleteKeyPair(ctx context.Context, keyPairID string) error

	// Load Balancer Management
	CreateLoadBalancer(ctx context.Context, req CreateLoadBalancerRequest) (*LoadBalancer, error)
	GetLoadBalancer(ctx context.Context, lbID string) (*LoadBalancer, error)
	ListLoadBalancers(ctx context.Context) ([]LoadBalancer, error)
	UpdateLoadBalancer(ctx context.Context, lbID string, req UpdateLoadBalancerRequest) (*LoadBalancer, error)
	DeleteLoadBalancer(ctx context.Context, lbID string) error
}

// VPC represents a Virtual Private Cloud
type VPC struct {
	ID        string            `json:"id"`
	Name      string            `json:"name"`
	CIDR      string            `json:"cidr"`
	State     string            `json:"state"`
	Region    string            `json:"region"`
	CreatedAt string            `json:"created_at"`
	Tags      map[string]string `json:"tags,omitempty"`
}

// Subnet represents a subnet within a VPC
type Subnet struct {
	ID               string            `json:"id"`
	Name             string            `json:"name"`
	VPCID            string            `json:"vpc_id"`
	CIDR             string            `json:"cidr"`
	AvailabilityZone string            `json:"availability_zone"`
	State            string            `json:"state"`
	CreatedAt        string            `json:"created_at"`
	Tags             map[string]string `json:"tags,omitempty"`
}

// SecurityGroup represents a security group
type SecurityGroup struct {
	ID          string            `json:"id"`
	Name        string            `json:"name"`
	Description string            `json:"description"`
	VPCID       string            `json:"vpc_id"`
	CreatedAt   string            `json:"created_at"`
	Tags        map[string]string `json:"tags,omitempty"`
}

// SecurityGroupRule represents a security group rule
type SecurityGroupRule struct {
	ID              string `json:"id"`
	SecurityGroupID string `json:"security_group_id"`
	Type            string `json:"type"`     // ingress, egress
	Protocol        string `json:"protocol"` // tcp, udp, icmp, all
	PortFrom        int    `json:"port_from"`
	PortTo          int    `json:"port_to"`
	Source          string `json:"source"` // CIDR or security group ID
	Description     string `json:"description"`
	CreatedAt       string `json:"created_at"`
}

// KeyPair represents a key pair for SSH access
type KeyPair struct {
	ID          string            `json:"id"`
	Name        string            `json:"name"`
	Fingerprint string            `json:"fingerprint"`
	PublicKey   string            `json:"public_key"`
	PrivateKey  string            `json:"private_key,omitempty"` // Only returned on creation
	CreatedAt   string            `json:"created_at"`
	Tags        map[string]string `json:"tags,omitempty"`
}

// LoadBalancer represents a load balancer
type LoadBalancer struct {
	ID        string            `json:"id"`
	Name      string            `json:"name"`
	Type      string            `json:"type"` // application, network, classic
	State     string            `json:"state"`
	VPCID     string            `json:"vpc_id"`
	SubnetIDs []string          `json:"subnet_ids"`
	DNSName   string            `json:"dns_name"`
	Port      int               `json:"port"`
	Protocol  string            `json:"protocol"`
	CreatedAt string            `json:"created_at"`
	Tags      map[string]string `json:"tags,omitempty"`
}

// Request types
type CreateVPCRequest struct {
	Name   string            `json:"name"`
	CIDR   string            `json:"cidr"`
	Region string            `json:"region"`
	Tags   map[string]string `json:"tags,omitempty"`
}

type UpdateVPCRequest struct {
	Name string            `json:"name,omitempty"`
	Tags map[string]string `json:"tags,omitempty"`
}

type CreateSubnetRequest struct {
	Name             string            `json:"name"`
	VPCID            string            `json:"vpc_id"`
	CIDR             string            `json:"cidr"`
	AvailabilityZone string            `json:"availability_zone"`
	Tags             map[string]string `json:"tags,omitempty"`
}

type UpdateSubnetRequest struct {
	Name string            `json:"name,omitempty"`
	Tags map[string]string `json:"tags,omitempty"`
}

type CreateSecurityGroupRequest struct {
	Name        string            `json:"name"`
	Description string            `json:"description"`
	VPCID       string            `json:"vpc_id"`
	Tags        map[string]string `json:"tags,omitempty"`
}

type UpdateSecurityGroupRequest struct {
	Name        string            `json:"name,omitempty"`
	Description string            `json:"description,omitempty"`
	Tags        map[string]string `json:"tags,omitempty"`
}

type CreateSecurityGroupRuleRequest struct {
	SecurityGroupID string `json:"security_group_id"`
	Type            string `json:"type"` // ingress, egress
	Protocol        string `json:"protocol"`
	PortFrom        int    `json:"port_from"`
	PortTo          int    `json:"port_to"`
	Source          string `json:"source"`
	Description     string `json:"description,omitempty"`
}

type CreateKeyPairRequest struct {
	Name      string            `json:"name"`
	PublicKey string            `json:"public_key,omitempty"` // If not provided, will be generated
	Tags      map[string]string `json:"tags,omitempty"`
}

type CreateLoadBalancerRequest struct {
	Name      string            `json:"name"`
	Type      string            `json:"type"`
	VPCID     string            `json:"vpc_id"`
	SubnetIDs []string          `json:"subnet_ids"`
	Port      int               `json:"port"`
	Protocol  string            `json:"protocol"`
	Tags      map[string]string `json:"tags,omitempty"`
}

type UpdateLoadBalancerRequest struct {
	Name string            `json:"name,omitempty"`
	Tags map[string]string `json:"tags,omitempty"`
}

File: frontend/src/services/network.ts
/**
 * Network Service
 * Network 관련 API 호출
 */

import { BaseService } from '@/lib/service-base';
import { API_ENDPOINTS } from '@/lib/api-endpoints';
import type {
  VPC,
  Subnet,
  SecurityGroup,
  CreateVPCForm,
  CreateSubnetForm,
  CreateSecurityGroupForm,
  SecurityGroupRule,
  CloudProvider,
} from '@/lib/types';

class NetworkService extends BaseService {
  // ===== VPC Management =====
  
  /**
   * VPC 목록 조회
   * 
   * @param provider - 클라우드 프로바이더 (aws, gcp, azure 등)
   * @param credentialId - 자격 증명 ID
   * @param region - 리전 (선택사항)
   * @returns VPC 배열
   * 
   * @example
   * ```tsx
   * const vpcs = await networkService.listVPCs('aws', 'credential-id', 'ap-northeast-2');
   * ```
   */
  async listVPCs(
    provider: CloudProvider,
    credentialId: string,
    region?: string
  ): Promise<VPC[]> {
    // 1. API 호출하여 VPC 목록 가져오기
    const data = await this.get<{ vpcs: VPC[] }>(
      API_ENDPOINTS.network.vpcs.list(provider, credentialId, region)
    );
    // 2. 응답 데이터에서 vpcs 배열 추출 (없으면 빈 배열)
    return data.vpcs || [];
  }

  /**
   * 특정 VPC 조회
   * 
   * @param provider - 클라우드 프로바이더
   * @param vpcId - VPC ID
   * @param credentialId - 자격 증명 ID
   * @param region - 리전
   * @returns VPC 정보
   * 
   * @example
   * ```tsx
   * const vpc = await networkService.getVPC('aws', 'vpc-123', 'credential-id', 'ap-northeast-2');
   * ```
   */
  async getVPC(
    provider: CloudProvider,
    vpcId: string,
    credentialId: string,
    region: string
  ): Promise<VPC> {
    return this.get<VPC>(
      API_ENDPOINTS.network.vpcs.detail(provider, vpcId, credentialId, region)
    );
  }

  /**
   * VPC 생성
   * 
   * @param provider - 클라우드 프로바이더
   * @param data - VPC 생성 데이터 (name, cidr_block, region 등)
   * @returns 생성된 VPC 정보
   * 
   * @example
   * ```tsx
   * const vpc = await networkService.createVPC('aws', {
   *   credential_id: 'credential-id',
   *   name: 'my-vpc',
   *   cidr_block: '10.0.0.0/16',
   *   region: 'ap-northeast-2',
   * });
   * ```
   */
  async createVPC(
    provider: CloudProvider,
    data: CreateVPCForm
  ): Promise<VPC> {
    return this.post<VPC>(
      API_ENDPOINTS.network.vpcs.create(provider),
      data
    );
  }

  async updateVPC(
    provider: CloudProvider,
    vpcId: string,
    data: Partial<CreateVPCForm>,
    credentialId: string,
    region: string
  ): Promise<VPC> {
    return this.put<VPC>(
      API_ENDPOINTS.network.vpcs.update(provider, vpcId, credentialId, region),
      data
    );
  }

  async deleteVPC(
    provider: CloudProvider,
    vpcId: string,
    credentialId: string,
    region: string
  ): Promise<void> {
    return this.delete<void>(
      API_ENDPOINTS.network.vpcs.delete(provider, vpcId, credentialId, region)
    );
  }

  // Subnet 관리
  async listSubnets(
    provider: CloudProvider,
    credentialId: string,
    vpcId: string,
    region: string
  ): Promise<Subnet[]> {
    const data = await this.get<{ subnets: Subnet[] }>(
      API_ENDPOINTS.network.subnets.list(provider, credentialId, vpcId, region)
    );
    return data.subnets || [];
  }

  async getSubnet(
    provider: CloudProvider,
    subnetId: string,
    credentialId: string,
    region: string
  ): Promise<Subnet> {
    return this.get<Subnet>(
      API_ENDPOINTS.network.subnets.detail(provider, subnetId, credentialId, region)
    );
  }

  async createSubnet(
    provider: CloudProvider,
    data: CreateSubnetForm
  ): Promise<Subnet> {
    return this.post<Subnet>(
      API_ENDPOINTS.network.subnets.create(provider),
      data
    );
  }

  async updateSubnet(
    provider: CloudProvider,
    subnetId: string,
    data: Partial<CreateSubnetForm>,
    credentialId: string,
    region: string
  ): Promise<Subnet> {
    return this.put<Subnet>(
      API_ENDPOINTS.network.subnets.update(provider, subnetId, credentialId, region),
      data
    );
  }

  async deleteSubnet(
    provider: CloudProvider,
    subnetId: string,
    credentialId: string,
    region: string
  ): Promise<void> {
    return this.delete<void>(
      API_ENDPOINTS.network.subnets.delete(provider, subnetId, credentialId, region)
    );
  }

  // Security Group 관리
  async listSecurityGroups(
    provider: CloudProvider,
    credentialId: string,
    vpcId: string,
    region: string
  ): Promise<SecurityGroup[]> {
    const data = await this.get<{ security_groups: SecurityGroup[] }>(
      API_ENDPOINTS.network.securityGroups.list(provider, credentialId, vpcId, region)
    );
    return data.security_groups || [];
  }

  async getSecurityGroup(
    provider: CloudProvider,
    securityGroupId: string,
    credentialId: string,
    region: string
  ): Promise<SecurityGroup> {
    return this.get<SecurityGroup>(
      API_ENDPOINTS.network.securityGroups.detail(provider, securityGroupId, credentialId, region)
    );
  }

  async createSecurityGroup(
    provider: CloudProvider,
    data: CreateSecurityGroupForm
  ): Promise<SecurityGroup> {
    return this.post<SecurityGroup>(
      API_ENDPOINTS.network.securityGroups.create(provider),
      data
    );
  }

  async updateSecurityGroup(
    provider: CloudProvider,
    securityGroupId: string,
    data: Partial<CreateSecurityGroupForm>,
    credentialId: string,
    region: string
  ): Promise<SecurityGroup> {
    return this.put<SecurityGroup>(
      API_ENDPOINTS.network.securityGroups.update(provider, securityGroupId, credentialId, region),
      data
    );
  }

  async deleteSecurityGroup(
    provider: CloudProvider,
    securityGroupId: string,
    credentialId: string,
    region: string
  ): Promise<void> {
    return this.delete<void>(
      API_ENDPOINTS.network.securityGroups.delete(provider, securityGroupId, credentialId, region)
    );
  }

  // Security Group Rule 관리
  async addSecurityGroupRule(
    provider: CloudProvider,
    securityGroupId: string,
    rule: SecurityGroupRule,
    credentialId: string,
    region: string
  ): Promise<SecurityGroup> {
    return this.post<SecurityGroup>(
      API_ENDPOINTS.network.securityGroups.rules.add(provider, securityGroupId, credentialId, region),
      rule
    );
  }

  async removeSecurityGroupRule(
    provider: CloudProvider,
    securityGroupId: string,
    ruleId: string,
    credentialId: string,
    region: string
  ): Promise<void> {
    return this.delete<void>(
      API_ENDPOINTS.network.securityGroups.rules.remove(provider, securityGroupId, credentialId, region, ruleId)
    );
  }

  async updateSecurityGroupRules(
    provider: CloudProvider,
    securityGroupId: string,
    rules: SecurityGroupRule[],
    credentialId: string,
    region: string
  ): Promise<SecurityGroup> {
    return this.put<SecurityGroup>(
      API_ENDPOINTS.network.securityGroups.rules.update(provider, securityGroupId, credentialId, region),
      { rules }
    );
  }
}

export const networkService = new NetworkService();
